src/types/mod.rs

```diff
@@ -0,0 +1,811 @@
+// Rust Simplicity Library
+// Written in 2023 by
+//   Andrew Poelstra <apoelstra@blockstream.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Types and Type Inference
+//!
+//! Every Simplicity expression has two types associated with it: a source and
+//! a target. These types are inferred from the structure of the program, with
+//! the additional rule that any unbound types in a complete program are simply
+//! set to the unit type.
+//!
+//! Concrete are one of three things
+//!   * A unit type, which has one value
+//!   * A sum of two other types
+//!   * A product of two other types
+//!
+//! During type inference, types may also be *free*, which means that they may
+//! turn out to be any type. Additional constraints are provided by the program
+//! itself, and type inference proceeds by using the union-bound algorithm to
+//! unify all the constraints on each type.
+//!
+//! If unification succeeds, types may then be finalized, which sets all free
+//! types to unit and does one additional check (the "occurs check") which
+//! makes sure there are no infinitely-sized types, which are illegal in
+//! Simplicity.
+//!
+//! There are four main types in this module:
+//!   * `Type`, which holds a mutex holding a `TypeInner`. This is the main
+//!     type representing a Simplicity type. This is the only type that users
+//!     are likely to encounter; the others are mainly only used during type
+//!     inference.
+//!   * `Constraint` defines whether a given type is free, bound to a
+//!     particular `Bound`, or equal to another variable.
+//!   * `Bound` defines the actual structure of a type: whether it is a
+//!     unit, sum or product.
+//!   * `Final`, which for a finalized type, provides a simple mutex-free
+//!     structure containing all of the type information and its TMR.
```

Comment:
I would like to eradicate the word "variable" and replace it with "type". Something like this: 

- `Type` holds a mutex to a `TypeInner` which in turn contains a `Constraint`. This is the main representation of a Simplicity type, and the only struct that users are likely going to encounter. `Type`s may be variable and incomplete.
- `Constraint` defines whether a `Type` is free, bound to a particular `Bound` or equal to another `Type`.
- `Bound` defines the structure of a `Type`: whether it is a unit, or the sum or product of two `Types`.
- `Final` is a mutex-free structure that can be obtained from a finalized `Type` and includes all its information and its TMR.